### PR TITLE
helper elements test cases for event emission to parent component

### PR DESCRIPTION
**Pull request by nitinkalenk**
I have added only three cases today..
Please review it extensively and comment...

### CODE REVIEW

File: projects/isteven-multiselect/src/lib/helper-elements/helper-elements.component.spec.ts [hunk 1]
```diff
@@ -16,22 +18,28 @@ describe('HelperElementsComponent', () => {
...
       // arrange
+      let selectAllBtnClicked = false;
+      component.selectAllClicked.subscribe(event => {
+        selectAllBtnClicked = true;
+      });
       // act
+      component.selectAll();
       // assert
-      expect(true).toBeTruthy();
+      expect(selectAllBtnClicked).toBe(true);
```
Comment (pankajparkar) [inline projects/isteven-multiselect/src/lib/helper-elements/helper-elements.component.spec.ts]: I'd rather recommend to test it by clicking element by code. By this way it ensures both the things, 

- like from UI perspective events are triggered correctly
- triggered event is calling underlying method or not? 

You can take a reference of the same from [this test case](https://github.com/pankajparkar/ngx-isteven-multiselect/blob/feature/test-cases/projects/isteven-multiselect/src/lib/grouped-options/grouped-options.component.spec.ts#L69-L84)

Comment (pankajparkar) [reply projects/isteven-multiselect/src/lib/helper-elements/helper-elements.component.spec.ts]: Same comment is applied for all other test cases.

**Comment by nitinkalenk** [comment]
completed test cases specified for helper elements and added one one new test case also...please review it rigirously

File: projects/isteven-multiselect/src/lib/helper-elements/helper-elements.component.spec.ts [hunk 2]
```diff
@@ -44,25 +62,45 @@ describe('HelperElementsComponent', () => {
...
       // arrange
+      component.multiple = true;
+      fixture.detectChanges();
+      let selectNoneBtnClicked = false;
+      component.selectNoneClicked.subscribe(event => {
+        selectNoneBtnClicked = true;
+      });
       // act
+      const selectNoneBtn = debugElement.queryAll(By.css('button.helper-button'))
+            .find(btn => btn.nativeElement.textContent.toLowerCase().indexOf('select none') !== -1);
```
Comment (pankajparkar) [inline projects/isteven-multiselect/src/lib/helper-elements/helper-elements.component.spec.ts]: Seems like we have repeated `btn.nativeElement.textContent.toLowerCase().indexOf('select none') !== -1)` somewhat this line of code multiple time, and we're going to use it at multiple places. Consider plucking it out from there.

Create a `specHelper.ts` file and write this generic functions over there and reuse it.

File: projects/isteven-multiselect/src/lib/helper-elements/helper-elements.component.spec.ts [hunk 3]
```diff
@@ -72,18 +110,49 @@ describe('HelperElementsComponent', () => {
   })
 
   describe('Select all and select none button should be', () => {
+
     it('visible only in case of multiple select', () => {
       // arrange
+      component.multiple = true;
+      fixture.detectChanges();
+      const buttons = debugElement.queryAll(By.css('button.helper-button'));
+      const selectAllBtn = buttons.filter(btn => btn.nativeElement.textContent.toLowerCase().indexOf('select all') !== -1);
```
Comment (pankajparkar) [inline projects/isteven-multiselect/src/lib/helper-elements/helper-elements.component.spec.ts]: We have repeated `select all` & `select none` as a string in multiple places, you can declare this variable as a constant like 

`const SELECT_ALL = 'select all'`

and use it wherever you need.

File: projects/isteven-multiselect/src/lib/helper-elements/helper-elements.component.spec.ts [hunk 4]
```diff
@@ -72,18 +110,49 @@ describe('HelperElementsComponent', () => {
...
+      expect(selectNoneBtn.length).toBe(1);
     })
+
     it('invisible incase of single select', () => {
       // arrange
+      component.multiple = false;
+      fixture.detectChanges();
+      const buttons = debugElement.queryAll(By.css('button.helper-button'));
+      const selectAllBtn = buttons.filter(btn => btn.nativeElement.textContent.toLowerCase().indexOf('select all') !== -1);
+      const selectNoneBtn = buttons.filter(btn => btn.nativeElement.textContent.toLowerCase().indexOf('select none') !== -1);
```
Comment (pankajparkar) [inline projects/isteven-multiselect/src/lib/helper-elements/helper-elements.component.spec.ts]: Spot on!!! I didn't know `buttons` will already have collection. I learnt new thing from this.
I'll undo few things in my test cases now :p

File: projects/isteven-multiselect/src/lib/helper-elements/helper-elements.component.spec.ts [hunk 5]
```diff
@@ -1,38 +1,59 @@
...
       // arrange
+      component.multiple = true;
+      fixture.detectChanges();
+      let selectAllBtnClicked = false;
+      component.selectAllClicked.subscribe(event => {
+        selectAllBtnClicked = true;
+      });
       // act
+      const selectAllBtn = debugElement.queryAll(By.css('button.helper-button'))
+            .find(btn => btn.nativeElement.textContent.toLowerCase().indexOf('select all') !== -1);
```
Comment (pankajparkar) [inline projects/isteven-multiselect/src/lib/helper-elements/helper-elements.component.spec.ts]: Couple of things here
 
1. Replace 'select all' with `SELECT_ALL` const.
2. use `checkForElement` function inside `.find` method.

File: projects/isteven-multiselect/src/lib/helper-elements/helper-elements.component.spec.ts [hunk 6]
```diff
@@ -72,18 +113,49 @@ describe('HelperElementsComponent', () => {
...
     })
   })
 
+  describe('Reset button should be', () => {
+    it('visible in case of multiple select and single select', () => {
+      // arrange
+      component.multiple = false;
+      fixture.detectChanges();
+      const buttons = debugElement.queryAll(By.css('button.helper-button'));
+      let resetBtn = buttons.filter(btn => btn => checkElementForText(btn, RESET));
```
Comment (pankajparkar) [inline projects/isteven-multiselect/src/lib/helper-elements/helper-elements.component.spec.ts]: why `btn => btn =>` two arrow function, are you trying to pass callback?
I believe it should be `btn =>`

File: projects/isteven-multiselect/src/lib/grouped-options/grouped-options.component.spec.ts [hunk 7]
```diff
@@ -60,21 +60,54 @@ describe('GroupedOptionsComponent', () => {
   describe('Group option', () => {
     it('on click on group options, it should select all underlying options', () => {
       // arrange
+      const group = debugElement.query(By.css('.group:first-child .option:first-child'))
```
Comment (pankajparkar) [inline projects/isteven-multiselect/src/lib/grouped-options/grouped-options.component.spec.ts]: Not completely sure, why this changes are being shown at your changes?
Can you please update your master repository and push it inside your local branch?

File: projects/isteven-multiselect/src/lib/helper-elements/helper-elements.component.spec.ts (hunk 6 above)
Comment (nitinkalenk) [reply projects/isteven-multiselect/src/lib/helper-elements/helper-elements.component.spec.ts]: my bad.. a typo actually...

File: projects/isteven-multiselect/src/lib/grouped-options/grouped-options.component.spec.ts (hunk 7 above)
Comment (nitinkalenk) [reply projects/isteven-multiselect/src/lib/grouped-options/grouped-options.component.spec.ts]: sorry about this made some mistake while pulling from repo
